### PR TITLE
Added 2 Ops: Floor divide scalar and Floor divide scalar mode

### PR DESCRIPTION
**Pull request by IanWood1**
- Added linalg lowering for `AtenFloorDivideScalarOp`
  - Needed `AtenDivScalarModeOp` for the decomp.
- Added linalg lowering for `AtenDivScalarModeOp`
  - Moved linalg payload logic to `createDivModePayload()` since the logic was nearly identical for both `AtenDivScalarModeOp` and `AtenDivTensorModeOp`. Just a template function 
 -  Added `AtenDivScalarModeOp` lowering for stablehlo
 

Pytorch's [`torch.floor_divide()`](https://pytorch.org/docs/stable/generated/torch.floor_divide.html) in a previous version (for a reason unknown to me) preformed a truncation instead of "floor". The already implemented op `AtenFloorDivideTensorOp` was done before this change.  However, this wasn't caught because our testcases only tested positive floor division. I changed this to floor as well as adding a few test cases.

### CODE REVIEW

File: lib/Conversion/TorchToLinalg/Uncategorized.cpp [hunk 1]
```diff
@@ -25,7 +25,10 @@
 #include "torch-mlir/Dialect/Torch/Utils/TorchUpstream.h"
 #include "torch-mlir/Dialect/Torch/Utils/Utils.h"
 #include "llvm/ADT/APSInt.h"
+#include <cassert>
+#include <iostream>
```
Comment (renxida) [inline lib/Conversion/TorchToLinalg/Uncategorized.cpp:29]: apparently including `<iostream>` is one of the cardinal sins in a LLVM project file.

Why is this included and can you do the same with just the llvm::errs etc streams?

**Comment by IanWood1** [comment]
nod-ai/SHARK-Turbine#618

File: lib/Conversion/TorchToLinalg/Uncategorized.cpp (hunk 1 above)
Comment (renxida) [reply lib/Conversion/TorchToLinalg/Uncategorized.cpp:29]: https://llvm.org/docs/CodingStandards.html#include-iostream-is-forbidden

Comment (renxida) [reply lib/Conversion/TorchToLinalg/Uncategorized.cpp:29]: (looking further it looks like you're moving some code from one file to another)

> 
> [Use raw_ostream](https://llvm.org/docs/CodingStandards.html#id52)
> LLVM includes a lightweight, simple, and efficient stream implementation in llvm/Support/raw_ostream.h, which provides all of the common features of std::ostream. All new code should use raw_ostream instead of ostream.
> 
> Unlike std::ostream, raw_ostream is not a template and can be forward declared as class raw_ostream. Public headers should generally not include the raw_ostream header, but use forward declarations and constant references to raw_ostream instances.
-- https://llvm.org/docs/CodingStandards.html#use-raw-ostream

Since we're already touching this code, let's modernize this part

File: lib/Conversion/TorchToLinalg/Uncategorized.cpp [hunk 2]
```diff
@@ -213,6 +216,79 @@ createTriangularMatrix(OpBuilder &b, Location loc, ValueRange payloadArgs,
   return success();
 }
 
+template <typename OpT>
+Value createDivModePayload(OpBuilder &b, Location loc,
+                           const TypeConverter *converter,
+                           ValueRange payloadArgs, OpT op,
+                           ArrayRef<Value> operands) {
+  static_assert(std::is_same_v<OpT, AtenDivTensorModeOp> ||
```
Comment (renxida) [inline lib/Conversion/TorchToLinalg/Uncategorized.cpp:224]: nice touch

File: lib/Conversion/TorchToStablehlo/Basic.cpp [hunk 3]
```diff
@@ -432,21 +432,25 @@ class ConvertAtenMulDivOp : public OpConversionPattern<AtenOpT> {
...
-    if (!matchPattern(divTensorModeOp.getRoundingMode(),
-                      m_TorchConstantStr(roundingMode)))
+    if (!matchPattern(opRoundingMode, m_TorchConstantStr(roundingMode))) {
       return rewriter.notifyMatchFailure(
           op, "only support constant str rounding mode");
+    }
 
     // if trunc and int, do nothing
-    if (roundingMode == "trunc" && isa<mlir::FloatType>(outElemTy)) {
+    if (roundingMode == "trunc" && outElemTy.isa<mlir::FloatType>()) {
```
Comment (renxida) [inline lib/Conversion/TorchToStablehlo/Basic.cpp:453]: @rsuderman you mentioned earlier that we want to write casts using the non-member function. Does that apply to `isa<>()`? Or does the opposite apply?

File: lib/Conversion/TorchToLinalg/Uncategorized.cpp (hunk 1 above)
Comment (IanWood1) [reply lib/Conversion/TorchToLinalg/Uncategorized.cpp:29]: embarrassing... it's just left over from some "debugging"

File: lib/Conversion/TorchToLinalg/Uncategorized.cpp [hunk 4]
```diff
@@ -25,7 +25,10 @@
 #include "torch-mlir/Dialect/Torch/Utils/TorchUpstream.h"
 #include "torch-mlir/Dialect/Torch/Utils/Utils.h"
 #include "llvm/ADT/APSInt.h"
+#include <cassert>
```
Comment (rsuderman) [inline lib/Conversion/TorchToLinalg/Uncategorized.cpp:28]: Do you still need `cassert`? I see your static assert below.

File: lib/Conversion/TorchToLinalg/Uncategorized.cpp [hunk 5]
```diff
@@ -213,6 +216,79 @@ createTriangularMatrix(OpBuilder &b, Location loc, ValueRange payloadArgs,
...
+Value createDivModePayload(OpBuilder &b, Location loc,
+                           const TypeConverter *converter,
+                           ValueRange payloadArgs, OpT op,
+                           ArrayRef<Value> operands) {
+  static_assert(std::is_same_v<OpT, AtenDivTensorModeOp> ||
+                    std::is_same_v<OpT, AtenDivScalarModeOp>,
+                "template type must be a tensor/scalar div mode");
+  typename OpT::Adaptor adaptor(operands);
+  Type dtype = converter->convertType(op.getType())
+                   .template cast<RankedTensorType>()
```
Comment (rsuderman) [inline lib/Conversion/TorchToLinalg/Uncategorized.cpp:229]: Typically we do `cast<RankedTensorType>(...)` instead of using the templated member function. It avoids the extract `template` part.

File: lib/Conversion/TorchToStablehlo/Basic.cpp (hunk 3 above)
Comment (rsuderman) [reply lib/Conversion/TorchToStablehlo/Basic.cpp:453]: I think we want to do the non-member version? IIRC it does a pointer case and some magic but its preferable due to better behavior for templated code. Similar comment at the top.

**Review by rsuderman — CHANGES_REQUESTED** [review]

File: lib/Conversion/TorchToStablehlo/Basic.cpp [hunk 6]
```diff
@@ -432,7 +432,8 @@ class ConvertAtenMulDivOp : public OpConversionPattern<AtenOpT> {
     Value result =
         rewriter.create<ChloOpT>(loc, outType, lhs, rhs, bcastDimensions);
 
-    if (!isa<AtenDivTensorModeOp, AtenDivScalarModeOp>(&op)) {
+    if (!std::is_same<AtenDivTensorModeOp, AtenOpT>() ||
```
Comment (IanWood1) [inline lib/Conversion/TorchToStablehlo/Basic.cpp]: Just changed isa to `is_same` since the type can be evaluated against tensor/scalar mode at compile time (if could be marked `constexpr` but idk how important that is)

**Comment by IanWood1** [comment]
made requested changes: use non-member functions and removed includes

**Review by renxida — APPROVED** [review]

**Review by rsuderman — APPROVED** [review]

**Comment by IanWood1** [comment]
closes #3143
